### PR TITLE
Report heading and compass point in the Measure tool (#1817)

### LINKED ISSUE #1817: [Feature]: Report heading/azimuth in the Measure tool
### Before you start

- [x] I searched the [existing issues](https://github.com/opengeos/GeoLibre/issues) and did not find a duplicate request.

### What problem does this solve?

The Measure tool reports distance and area, but no direction. The upstream `Measurement` interface (`maplibre-gl-components`, `dist/types/lib/core/types.d.ts`) carries `distance`, `area`, and `segments` — there is no bearing field.

Google Earth Pro's ruler reports a heading alongside the distance. Beyond being useful in its own right (aspect, alignment, sight lines, "which way does this runway point"), it doubles as a **correctness check**: instructional material routinely says "your heading should be about 310 degrees", which tells a student immediately whether they drew the line in the right direction or clicked the two points in the wrong order.

Found while auditing an undergraduate textbook's GEP labs against GeoLibre — two labs reference the heading readout.

### Proposed solution

No upstream change needed. `packages/plugins/src/plugins/terrain-measure.ts` already listens to the Measure control's measurement events and appends its own **Terrain (3D)** section to the control's panel. Add azimuth in the same place:

- for a two-point line, the forward azimuth from first point to last
- for a multi-point path, per-segment azimuths, plus the overall first-to-last azimuth

Report as degrees from true north. Worth deciding whether to show initial bearing, final bearing, or both for long geodesics — over a short measurement they are identical, but on a trans-oceanic line they differ substantially, which is itself a nice thing to be able to show.

Magnetic bearing is a possible follow-up but needs a declination model; true north only for a first pass.

### Estimated scope

Small: a minor feature or UI tweak

### Alternatives considered

Computing it in the expression builder from the measured geometry. Possible, but absurdly indirect for a number the tool is one line of trigonometry away from already having.

### PR DESCRIPTION
**Pull request by giswqs**
Closes #1817.

The Measure panel reported distance and area but no direction. The upstream `Measurement` interface carries `distance`, `area`, and `segments` and no bearing, so lab material that checks a student's work by heading ("your heading should be about 310 degrees") had nothing to read.

No upstream change needed, as the issue predicted: `terrain-measure.ts` already augments the Measure panel, so the bearing rows go in beside the existing terrain section.

## Its own section, not a row inside "Terrain (3D)"

This is the design call worth reviewing. A bearing is **pure geometry on the measured coordinates** — it needs no elevation at all. Putting it inside the terrain section would make it vanish exactly when the terrain readout gives up: off-Earth, no DEM tiles loaded, no network. Those are the cases where you still very much want to know which way the line points.

It also renders **synchronously**, so it appears the instant the line is drawn rather than after the elevation round-trip that the terrain section waits on.

## A true bearing, not a screen angle

The azimuth is a great-circle initial bearing. On a Mercator map, Washington D.C. → London *looks* due east; the real initial bearing is ~50°. There is an explicit test for this, because a naive `atan2(dy, dx)` on projected coordinates would pass every cardinal-direction test and then be quietly wrong on every real measurement.

Three presentation decisions:

- **Final bearing only when it differs** by ≥1° from the initial. A great circle's heading changes as you travel it, but on a short line the two ends agree — showing "310° / 310°" everywhere would be noise. Long geodesics show the convergence.
- **Multi-point paths report the overall first-to-last bearing**, not one row per segment. The issue suggested per-segment; a twenty-point path would bury the panel, and GEP's own Path tab shows no heading at all. Happy to add per-segment behind a disclosure if you'd rather.
- **A 16-point compass label** accompanies the degrees — `310° NW` is much easier to sanity-check against the map than the number alone.

## Lifecycle

The bearing section tracks which measurement it describes, so removing *that* measurement clears it while removing a different one leaves it alone; `clear` and detach tear it down alongside the terrain section. The row-building DOM helper is now shared between the two sections rather than duplicated.

## Tests

`tests/measure-heading.test.ts`, 13 cases: cardinals, `[0, 360)` normalisation including an antimeridian crossing, the great-circle-vs-flat-map check, final-bearing convergence, compass labels and their wrap, and the `bearingRows` gating (area measurements, single points, and a degenerate line whose ends coincide).

Full suite: 5635 pass, 0 fail. `npm run typecheck` and scoped `pre-commit` clean.

## i18n

`terrainMeasure.heading` / `finalHeading` in all 18 catalogs, using each language's navigation term — `Cap` (fr), `Азимут` (ru), `方位角` (zh), `Kurs` (de).

## Summary by CodeRabbit

- **New Features**
  - Terrain measurements now display initial and final compass headings for line measurements.
  - Supports 16 compass directions with great-circle calculations, including multi-point routes.
  - Heading results update or clear appropriately as measurements change.
  - Handles incomplete, degenerate, antipodal, and unsupported measurements gracefully.

- **Localization**
  - Added translated heading labels across all supported languages.

- **Tests**
  - Added coverage for heading calculations, compass directions, multi-point measurements, and edge cases.

### CODE REVIEW

File: packages/plugins/src/plugins/terrain-measure-geometry.ts [hunk 1]
```diff
@@ -346,3 +346,59 @@ function gradient(
...
+  return (degrees + 360) % 360;
+}
+
+/**
+ * Final great-circle azimuth arriving at `b` from `a`, in degrees clockwise
+ * from true north. Derived by reversing the leg and flipping 180°.
+ */
+export function finalAzimuthDegrees(a: LngLat, b: LngLat): number {
+  return (forwardAzimuthDegrees(b, a) + 180) % 360;
+}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure-geometry.ts:375]: Done in b49188ed — correct, and worth fixing properly: `isAntipodal` now short-circuits `bearingRows` to no rows. Writing it also caught a bug in my first attempt — I normalised the longitude separation with the signed `+540` wrap, which maps a true 180-degree separation to 180 rather than 0, so the guard never fired. It now normalises into [0, 360) and compares against 180, which also handles pairs across the antimeridian. Tests cover both the equatorial and an oblique antipodal pair.

File: packages/plugins/src/plugins/terrain-measure.ts [hunk 2]
```diff
@@ -320,6 +325,46 @@ export function terrainReadoutRows(
...
+  const a: LngLat = [first.lng, first.lat];
+  const b: LngLat = [last.lng, last.lat];
+  if (a[0] === b[0] && a[1] === b[1]) return [];
+
+  const initial = forwardAzimuthDegrees(a, b);
+  const rows: Array<[string, string]> = [
+    [terrainMeasureLabels.heading, `${initial.toFixed(0)}\u00b0 ${compassPoint(initial)}`],
+  ];
+  const final = finalAzimuthDegrees(a, b);
+  if (Math.abs(((final - initial + 540) % 360) - 180) >= 1) {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure.ts]: Done in b49188ed — both fixed, thank you. The final-heading gate now compares `toFixed(0)` output rather than raw degrees, so it can never add a row that renders identically to the first. And the degenerate check uses an epsilon instead of exact equality, so endpoints separated only by floating-point dust report no bearing rather than an arbitrary one. Tests for both.

File: packages/plugins/src/plugins/terrain-measure-geometry.ts [hunk 3]
```diff
@@ -346,3 +346,89 @@ function gradient(
...
+/**
+ * Whether two points are close enough that the bearing between them is noise.
+ *
+ * An exact coordinate comparison misses points that differ only by
+ * floating-point dust, which still produce a mathematically valid but
+ * essentially arbitrary heading.
+ */
+export function isDegenerateSegment(a: LngLat, b: LngLat, toleranceDegrees = 1e-9): boolean {
+  return Math.abs(a[0] - b[0]) <= toleranceDegrees && Math.abs(a[1] - b[1]) <= toleranceDegrees;
+}
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure-geometry.ts:409]: Done in 9889537a — real gap, and writing it caught a second mistake of mine: I had the comparison inverted (testing the wrapped gap against 180 rather than 0), so the guard never fired. Now wraps the longitude delta onto the circle and compares against zero, with a test for [180,5] vs [-180,5].

File: tests/measure-heading.test.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    for (const [lng, lat] of [
+      [0.4, 0.1],
+      [1, 0.05],
+      [-0.6, 0.2],
+    ] as [number, number][]) {
+      const rows = bearingRows({ mode: "distance", points: [pt(0, 0), pt(lng, lat)] });
+      if (rows.length === 2) {
+        assert.notEqual(rows[0][1], rows[1][1], `duplicate heading rows for ${lng},${lat}`);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/measure-heading.test.ts:153]: Done in 9889537a — you're right that it was vacuous. Asserts `rows.length === 1` for each fixture directly.

File: tests/measure-heading.test.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  });
+
+  it("stays empty for endpoints separated only by floating-point dust", () => {
+    const rows = bearingRows({
+      mode: "distance",
+      points: [pt(5, 5), pt(5 + Number.EPSILON, 5 - Number.EPSILON)],
+    });
+    assert.deepEqual(rows, []);
+  });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/measure-heading.test.ts:167]: Done in 9889537a — good catch, `5 + Number.EPSILON === 5` so the fixture was only re-testing exact equality. Switched to 1e-12, with an assertion in the test that it really is a distinct double so the fixture cannot silently degrade again.

File: packages/plugins/src/plugins/terrain-measure-geometry.ts [hunk 6]
```diff
@@ -346,3 +346,89 @@ function gradient(
...
+ * from true north, normalised to [0, 360).
+ *
+ * This is the *initial* bearing: on a sphere a great-circle path's heading
+ * changes as you travel it, so a long line's arrival bearing differs from its
+ * departure bearing. {@link finalAzimuthDegrees} gives the other end.
+ *
+ * Radius-independent, so it needs no ellipsoid — the bearing between two
+ * coordinates is the same on any sphere.
+ */
+export function forwardAzimuthDegrees(a: LngLat, b: LngLat): number {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure-geometry.ts:362]: Done in 9889537a — same formula, so `forwardAzimuthDegrees` now delegates to `bearingBetween`. One implementation between the Measure heading and the Route Animation camera.

File: packages/plugins/src/plugins/terrain-measure.ts [hunk 7]
```diff
@@ -320,6 +327,51 @@ export function terrainReadoutRows(
...
+
+  const initial = forwardAzimuthDegrees(a, b);
+  const rows: Array<[string, string]> = [
+    [terrainMeasureLabels.heading, `${initial.toFixed(0)}\u00b0 ${compassPoint(initial)}`],
+  ];
+  // Compared on the *rendered* values, not the raw ones: a raw gap of just over
+  // a degree either side of a rounding boundary would otherwise add a second row
+  // showing the same number as the first.
+  const final = finalAzimuthDegrees(a, b);
+  if (final.toFixed(0) !== initial.toFixed(0)) {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure.ts]: Done in 9889537a — and you identified exactly the failure I traded for the other one. The gate now requires *both* a true convergence of at least a degree and distinct rendered values, so neither 45.4/45.6 (real gap too small) nor 89.5/90.4 (renders identically) produces a row.

File: packages/plugins/src/plugins/terrain-measure.ts [hunk 8]
```diff
@@ -320,6 +327,53 @@ export function terrainReadoutRows(
...
+  const last = measurement.points[measurement.points.length - 1];
+  const a: LngLat = [first.lng, first.lat];
+  const b: LngLat = [last.lng, last.lat];
+  // Coincident endpoints have no direction, and antipodal ones lie on
+  // infinitely many great circles, so neither has a bearing worth reporting.
+  if (isDegenerateSegment(a, b) || isAntipodal(a, b)) return [];
+
+  const initial = forwardAzimuthDegrees(a, b);
+  const rows: Array<[string, string]> = [
+    [terrainMeasureLabels.heading, `${initial.toFixed(0)}\u00b0 ${compassPoint(initial)}`],
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure.ts]: Done in 18458b78 — real bug, thank you. Both rows now render through one helper that rounds modulo 360, so 359.6 shows as `0° N` rather than `360° N`. Routing both through the same helper also keeps the duplicate-row gate comparing exactly what is displayed. Test added.

File: packages/plugins/src/plugins/terrain-measure.ts [hunk 9]
```diff
@@ -320,6 +327,53 @@ export function terrainReadoutRows(
...
+  // Both conditions are needed, and they guard opposite failures: the rendered
+  // values must differ (or the second row repeats the first verbatim), *and*
+  // the true convergence must be at least a degree (or bearings 0.2 apart that
+  // straddle a rounding boundary, say 45.4 and 45.6, earn a row for nothing).
+  const final = finalAzimuthDegrees(a, b);
+  const convergence = Math.abs(((final - initial + 540) % 360) - 180);
+  if (convergence >= 1 && final.toFixed(0) !== initial.toFixed(0)) {
+    rows.push([
+      terrainMeasureLabels.finalHeading,
+      `${final.toFixed(0)}\u00b0 ${compassPoint(final)}`,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/terrain-measure.ts]: Done in 18458b78 — real bug, thank you. Both rows now render through one helper that rounds modulo 360, so 359.6 shows as `0° N` rather than `360° N`. Routing both through the same helper also keeps the duplicate-row gate comparing exactly what is displayed. Test added.

File: apps/geolibre-desktop/src/i18n/locales/fa.json [hunk 10]
```diff
@@ -184,7 +184,9 @@
     "elevationRange": "کمینه / بیشینهٔ ارتفاع",
     "meanSlope": "شیب میانگین",
     "computing": "در حال محاسبهٔ زمین…",
-    "partialData": "برخی نمونه‌ها دادهٔ زمین نداشتند"
+    "partialData": "برخی نمونه‌ها دادهٔ زمین نداشتند",
+    "heading": "راستا",
+    "finalHeading": "راستای نهایی"
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/i18n/locales/fa.json]: Done in b7a8f8ad — verified and applied. `سمت` is indeed this catalog's established bearing term (`toolbar.setView.bearing`, `resetBearing`, `resetPitchBearing`), and `راستا` read as "alignment", so the labels are now `سمت` / `سمت نهایی`.

File: apps/geolibre-desktop/src/i18n/locales/it.json [hunk 11]
```diff
@@ -184,7 +184,9 @@
     "elevationRange": "Quota min / max",
     "meanSlope": "Pendenza media",
     "computing": "Calcolo del terreno…",
-    "partialData": "Alcuni campioni non avevano dati del terreno"
+    "partialData": "Alcuni campioni non avevano dati del terreno",
+    "heading": "Rotta",
+    "finalHeading": "Rotta finale"
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/i18n/locales/it.json]: Done in b7a8f8ad — took your `Azimut` / `Azimut finale` suggestion rather than `Orientamento`. `Rotta` is a route/course as you say, and `Orientamento` is already this catalog's term for the *camera* bearing (`toolbar.setView.bearing`, `printLayout.orientation`), so reusing it would have conflated the camera's orientation with a measured ground azimuth. `Azimut` keeps the two distinct.
